libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -955,23 +991,34 @@ std::string IRGenerator::deployCode(ContractDefinition const& _contract)
 	if (_contract.isLibrary())
 	{
 		solAssert(ContractType(_contract).immutableVariables().empty(), "");
-		immutables.emplace_back(std::map<std::string, std::string>{
-			{"immutableName"s, IRNames::libraryAddressImmutable()},
-			{"value"s, "address()"}
-		});
-
+		if (!eof)
+			immutables.emplace_back(std::map<std::string, std::string>{
+				{"immutableName"s, IRNames::libraryAddressImmutable()},
+				{"value"s, "address()"}
+			});
+		else
+			t("libraryAddressImmutableOffset", std::to_string(m_context.libraryAddressImmutableOffset()));
 	}
 	else
+	{
 		for (VariableDeclaration const* immutable: ContractType(_contract).immutableVariables())
 		{
 			solUnimplementedAssert(immutable->type()->isValueType());
 			solUnimplementedAssert(immutable->type()->sizeOnStack() == 1);
-			immutables.emplace_back(std::map<std::string, std::string>{
-				{"immutableName"s, std::to_string(immutable->id())},
-				{"value"s, "mload(" + std::to_string(m_context.immutableMemoryOffset(*immutable)) + ")"}
-			});
+			if (!eof)
+				immutables.emplace_back(std::map<std::string, std::string>{
+					{"immutableName"s, std::to_string(immutable->id())},
+					{"value"s, "mload(" + std::to_string(m_context.immutableMemoryOffset(*immutable)) + ")"}
+				});
 		}
+
+		t("auxDataStart", std::to_string(CompilerUtils::generalPurposeMemoryStart));
+		// TODO: Make sure that we cannot use reservedMemory() here
+		t("auxDataSize", std::to_string(m_context.reservedMemorySize()));
```

Comment:
One more thing. We should report an error when we know that the size of reserved memory exceeds the maximum size of auxdata.

This should be detected already in analysis, before we even proceed to code generation. [`ImmutableValidator`](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/ImmutableValidator.h) looks like a good place for it. Then here we should just have an assert against it. We already have a check in `Assembly.cpp` but that should be the last resort. Analysis has access to error reporter and can add useful info like the location of the error (we could point at the first immutable that goes over the limit).

It would also be fine to add just the assert for now and leave a TODO to add a proper analysis check later (but then mention it in #15310 so that we don't forget).

Reply:
I will add the assert but I'm wondering what happens if we have not only immutables in EOF data section. There can be also data and auxdata (different than EOF section aux data) which also goes to EOF data section.

Reply:
Assuming that you assert about `reservedMemorySize()` and not specifically about immutables, then the assert will still be correct. Any extra auxdata will still have to be passed in the same memory block, right?